presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -381,12 +386,68 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
 
     private static SetMultimap<Integer, ScheduledSplit> assignSourceDistributionSplits(Session session, List<ScheduledSplit> splits)
     {
-        int taskCount = getSparkInitialPartitionCount(session);
-        checkArgument(taskCount > 0, "taskCount must be greater then zero: %s", taskCount);
         ImmutableSetMultimap.Builder<Integer, ScheduledSplit> result = ImmutableSetMultimap.builder();
-        for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
-            result.put(splitIndex % taskCount, splits.get(splitIndex));
+
+        long maxSplitsSizeInBytesPerPartition = getMaxSplitsDataSizePerSparkPartition(session).toBytes();
+        checkArgument(maxSplitsSizeInBytesPerPartition > 0,
+                "maxSplitsSizeInBytesPerPartition must be greater than zero: %s", maxSplitsSizeInBytesPerPartition);
+
+        boolean splitsDataSizeAvailable = splits.stream()
+                .allMatch(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().isPresent());
+        boolean autoTuneInitialPartitionCount = isSparkInitialPartitionCountAutoTuneEnabled(session) && splitsDataSizeAvailable;
+
+        int initialPartitionCount;
+        if (!autoTuneInitialPartitionCount) {
+            initialPartitionCount = getSparkInitialPartitionCount(session);
+            checkArgument(initialPartitionCount > 0, "initialPartitionCount must be greater then zero: %s", initialPartitionCount);
+        }
+        else {
+            long totalSizeInBytes = splits.stream()
+                    .mapToLong(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong())
+                    .sum();
+            initialPartitionCount = max(1, toIntExact(totalSizeInBytes + maxSplitsSizeInBytesPerPartition - 1 / maxSplitsSizeInBytesPerPartition));
+        }
+
+        if (!splitsDataSizeAvailable) {
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                result.put(splitIndex % initialPartitionCount, splits.get(splitIndex));
+            }
+        }
+        else {
+            splits.sort((ScheduledSplit o1, ScheduledSplit o2) ->
+                    o1.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() >
```

Comment:
You can simply do `o1.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() - o2.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong()` (or the vice versa, depending on the order you are trying to achieve)

Reply:
Nevermind. DIdn't realize we are dealing with a long.